core/src/main/java/io/grpc/internal/DelayedClientTransport.java

```diff
@@ -359,22 +368,94 @@ public void run() {
+   * the stream or ping if the pick is successful, otherwise keep it pending.
+   *
+   * <p>If new pending streams are created while reprocess() is in progress, there is no guarantee
+   * that the pending streams will or will not be processed by this picker.
+   */
+  final void reprocess(LoadBalancer2.SubchannelPicker picker) {
+    ArrayList<PendingStream> toProcess;
+    ArrayList<PendingStream> toRemove = new ArrayList<PendingStream>();
+    synchronized (lock) {
+      if (pendingStreams == null || pendingStreams.isEmpty()) {
+        return;
+      }
+      toProcess = new ArrayList<PendingStream>(pendingStreams);
+    }
+
+    for (final PendingStream stream : toProcess) {
+      PickResult pickResult = picker.pickSubchannel(
+          stream.callOptions.getAffinity(), stream.headers);
+      final ClientTransport realTransport;
+      Subchannel subchannel = pickResult.getSubchannel();
+      if (subchannel != null) {
+        realTransport = obtainActiveTransport(subchannel);
+      } else {
+        realTransport = null;
+      }
+      if (realTransport != null) {
+        Executor executor = streamCreationExecutor;
+        // createRealStream may be expensive. It will start real streams on the transport. If
+        // there are pending requests, they will be serialized too, which may be expensive. Since
+        // we are now on transport thread, we need to offload the work to an executor.
+        if (stream.callOptions.getExecutor() != null) {
+          executor = stream.callOptions.getExecutor();
+        }
+        executor.execute(new Runnable() {
+            @Override
+            public void run() {
+              stream.createRealStream(realTransport);
+            }
+          });
+        toRemove.add(stream);
+      } else if (!pickResult.getStatus().isOk() && !stream.callOptions.isWaitForReady()) {
+        stream.setStream(new FailingClientStream(pickResult.getStatus()));
+        toRemove.add(stream);
+      }  // other cases: stay pending
+    }
+
+    synchronized (lock) {
+      if (pendingStreams == null || pendingStreams.isEmpty()) {
+        return;
+      }
+      pendingStreams.removeAll(toRemove);
+      if (pendingStreams.isEmpty()) {
+        pendingStreams = null;
+        // There may be a brief gap between delayed transport clearing in-use state, and first real
+        // transport starting streams and setting in-use state.  During the gap the whole channel's
+        // in-use state may be false. However, it shouldn't cause spurious switching to idleness
+        // (which would shutdown the transports and LoadBalancer) because the gap should be shorter
+        // than IDLE_MODE_DEFAULT_TIMEOUT_MILLIS (1 second).
+        listener.transportInUse(false);
+      }
+    }
+  }
+
+  // To be implemented by ManagedChannelImpl2 and unit test
+  @Nullable
+  ClientTransport obtainActiveTransport(Subchannel subchannel) {
+    throw new UnsupportedOperationException("Not implemented");
```

Comment:
Fair enough. I introduced a `SubchannelImpl` that adds `obtainActiveTransport()` to `Subchannel`.